File: dashboard/src/features/client-profiles/client-setup.ts
import { useQuery } from '@tanstack/react-query'
import { api } from '@/lib/api-client'
import { isMockMode } from '@/lib/mock-mode'
import type { DataClassification, HybridMode } from '@/types'

export interface ClientSetupCheck {
  apiKeyId: string
  model: string
  status: 'ready' | 'blocked'
  code: string
  checkedAtMs: number
  upstreamVerified: false
  defaultClassification?: DataClassification
  effectiveMode?: HybridMode
  ipRestricted?: boolean
  project?: { organizationId: string; projectId: string; environmentId: string }
}

export function useClientSetup(apiKeyId: string, model: string) {
  return useQuery({
    queryKey: ['client-setup', apiKeyId, model],
    queryFn: (): Promise<ClientSetupCheck> => isMockMode
      ? Promise.resolve({ apiKeyId, model, status: 'blocked', code: 'unverified', checkedAtMs: Date.now(), upstreamVerified: false })
      : api.get(`/admin/api-keys/${encodeURIComponent(apiKeyId)}/setup?model=${encodeURIComponent(model)}`),
    enabled: Boolean(apiKeyId && model),
    staleTime: 0,
    refetchOnMount: 'always',
    refetchOnWindowFocus: 'always',
    refetchInterval: 10_000,
  })
}

export function setupAllowsCopy(check: ClientSetupCheck | undefined, apiKeyId: string, model: string) {
  return Boolean(apiKeyId && model && check?.apiKeyId === apiKeyId && check.model === model && check.status === 'ready' && check.code === 'ready')
}

export function setupMessage(code?: string) {
  const messages: Record<string, string> = {
    ready: '所选密钥的模型权限、Provider 凭证和默认外发策略检查通过。',
    key_unavailable: '密钥已失效、被撤销或所属团队不可用。请检查密钥和团队状态。',
    control_only_key: '这是运维 Agent 控制密钥，不能调用模型。请选择独立的推理密钥。',
    owner_unavailable: '密钥所属用户不可用。请联系管理员恢复用户状态。',
    model_not_allowed: '密钥或团队未允许这个模型。请选择其他模型或联系管理员调整范围。',
    project_policy_denied: '项目外发策略未允许这条路由。请联系管理员检查 Provider、模型和外发边界。',
    local_only: '当前默认策略只允许本地执行，这个模型位于云端。请选择本地模型，或为已批准的项目明确数据分类。',
    model_unavailable: '模型已停用、未登记或没有符合策略的路由。请检查模型接入。',
    missing_credential: 'Provider 凭证未就绪。请联系管理员配置凭证后重试。',
    unverified: '演示模式无法验证真实接入，配置复制已禁用。',
  }
  return messages[code ?? ''] ?? '尚未获得有效的接入检查结果，配置复制已禁用。'
}

File: src/routes/client_setup.rs
use axum::{
    Json,
    extract::{Path, Query, State},
    http::HeaderMap,
};
use serde::Deserialize;
use serde_json::{Value, json};

use crate::{
    config::ResolvedProvider,
    control::UserCatalogGrant,
    governance::{HybridMode, ProjectPolicy, ProviderBoundary},
};

use super::{
    AppError, AppState, catalog_owner, effective_config, now_millis, require_console_user,
};

#[derive(Deserialize)]
pub(super) struct SetupQuery {
    model: String,
}

/// Read-only configuration preflight for a specific key. It deliberately does
/// not contact an upstream, reserve budget, or claim to validate the caller's
/// IP, payload capabilities, live health, or remaining quota.
pub(super) async fn check(
    State(state): State<AppState>,
    Path(key_id): Path<String>,
    Query(query): Query<SetupQuery>,
    headers: HeaderMap,
) -> Result<Json<Value>, AppError> {
    let actor = require_console_user(&state, &headers)?;
    let owner = catalog_owner(&state, &actor, Some(&key_id))?;
    if query.model.trim().is_empty() || query.model.len() > 256 {
        return Err(AppError::InvalidRequest(
            "model must contain 1–256 bytes".to_owned(),
        ));
    }
    let mut result = json!({
        "apiKeyId": key_id,
        "model": query.model,
        "status": "blocked",
        "code": "key_unavailable",
        "checkedAtMs": now_millis(),
        "upstreamVerified": false,
    });
    if state
        .auth
        .user_by_id(&owner)
        .is_none_or(|user| user.status != "active")
    {
        result["code"] = json!("owner_unavailable");
        return Ok(Json(result));
    }
    let grants = state.control.user_catalog_grants(&owner, Some(&key_id));
    let Some(grant) = grants.first() else {
        return Ok(Json(result));
    };
    if super::ensure_inference_purpose(grant.purpose.as_deref()).is_err() {
        result["code"] = json!("control_only_key");
        return Ok(Json(result));
    }
    let policy = state.governance.effective_policy(&grant.tenant);
    let mode = policy.effective_mode(None, policy.default_classification)?;
    result["project"] = json!({
        "organizationId": grant.tenant.organization_id.to_string(),
        "projectId": grant.tenant.project_id.to_string(),
        "environmentId": grant.tenant.environment_id.to_string(),
    });
    result["defaultClassification"] = json!(policy.default_classification);
    result["effectiveMode"] = json!(mode);
    result["ipRestricted"] = json!(grant.policy.ip_restricted);

    let config = effective_config(&state);
    let candidates = if config.smart_route_group(&query.model).is_some() {
        super::provider_view::catalog_smart_alias_candidates(
            &state,
            &config,
            &query.model,
            Some(&grant.policy),
            &policy,
        )
    } else {
        config.resolve(&query.model).into_iter().collect()
    };
    let mut code = "model_unavailable";
    for resolved in candidates {
        code = candidate_status(&state, grant, &policy, mode, &query.model, &resolved);
        if code == "ready" {
            break;
        }
    }
    result["code"] = json!(code);
    if code == "ready" {
        result["status"] = json!("ready");
    }
    Ok(Json(result))
}

fn candidate_status(
    state: &AppState,
    grant: &UserCatalogGrant,
    policy: &ProjectPolicy,
    mode: HybridMode,
    requested: &str,
    resolved: &ResolvedProvider,
) -> &'static str {
    if grant
        .policy
        .enforce_route(requested, &resolved.model, &resolved.provider_id)
        .is_err()
    {
        return "model_not_allowed";
    }
    if policy.enforce_attempt(resolved).is_err() {
        return "project_policy_denied";
    }
    if mode == HybridMode::LocalStrict
        && ProviderBoundary::for_resolved(resolved) == ProviderBoundary::Cloud
    {
        return "local_only";
    }
    if !resolved.provider.models.contains(&resolved.model) {
        return "model_unavailable";
    }
    let credential_ready = state
        .control
        .provider_credential_route_available(&resolved.provider_id)
        .unwrap_or_else(|| {
            !resolved.provider.api_key_required
                || resolved.provider.api_key().ok().flatten().is_some()
        });
    if !credential_ready {
        return "missing_credential";
    }
    "ready"
}
